src/bootstrap/src/core/build_steps/run.rs

```diff
@@ -118,7 +118,15 @@ impl Step for Miri {
     fn run(self, builder: &Builder<'_>) {
         let host = builder.build.build;
         let target = self.target;
-        let stage = builder.top_stage;
+
+        // `x run` uses stage 0 by default but miri does not work well with stage 0.
+        // Change the stage to 1 if it's not set explicitly.
+        let stage = if builder.config.is_explicit_stage() || builder.top_stage >= 1 {
+            builder.top_stage
+        } else {
+            1
+        };
```

Comment:
Seems a bit odd to use stage 1 if `--stage 0` is explicitly passed -- which this patch seems to do?

Reply:
An explicitly stage (even if it's 0) overrides the default, which is the expectation for most (if not all) people. And if you explicitly set stage 0, it will trigger a hard error a few lines below.